dcos node: Show masters

msabramo opened the pull request:
@tamarrow: This is the idea you suggested as a replacement for https://github.com/dcos/dcos-cli/pull/921

e.g.:

```
$ dcos node
   HOSTNAME         IP                          ID                    TYPE
 10.10.12.174  10.10.12.174  74df965e-a46e-46a9-8d2a-639db0af0e25-S0  slave
 10.10.12.183  10.10.12.183  74df965e-a46e-46a9-8d2a-639db0af0e25-S4  slave
 10.10.12.29   10.10.12.29   74df965e-a46e-46a9-8d2a-639db0af0e25-S2  slave
 10.10.3.201   10.10.3.201   74df965e-a46e-46a9-8d2a-639db0af0e25-S3  slave
master.mesos.  10.10.11.179                    N/A                    master
master.mesos.  10.10.12.68                     N/A                    master (leader)
master.mesos.  10.10.13.72                     N/A                    master
```

```
$ dcos node --field=version --field=pid
   HOSTNAME         IP                          ID                    TYPE             VERSION             PID
 10.10.12.174  10.10.12.174  74df965e-a46e-46a9-8d2a-639db0af0e25-S0  slave             1.0.1   slave(1)@10.10.12.174:5051
 10.10.12.183  10.10.12.183  74df965e-a46e-46a9-8d2a-639db0af0e25-S4  slave             1.0.1   slave(1)@10.10.12.183:5051
 10.10.12.29   10.10.12.29   74df965e-a46e-46a9-8d2a-639db0af0e25-S2  slave             1.0.1   slave(1)@10.10.12.29:5051
 10.10.3.201   10.10.3.201   74df965e-a46e-46a9-8d2a-639db0af0e25-S3  slave             1.0.1   slave(1)@10.10.3.201:5051
master.mesos.  10.10.11.179                    N/A                    master             N/A               N/A
master.mesos.  10.10.12.68     74df965e-a46e-46a9-8d2a-639db0af0e25   master (leader)   1.0.1    master@10.10.12.68:5050
master.mesos.  10.10.13.72                     N/A                    master             N/A               N/A
```

```
$ dcos node --json
[
  {
    "host": "master.mesos.",
    "ip": "10.10.11.179",
    "type": "master"
  },
  {
    "host": "master.mesos.",
    "ip": "10.10.13.72",
    "type": "master"
  },
  {
    "host": "master.mesos.",
    "ip": "10.10.12.68",
    "type": "master"
  },
  {
    "TASK_ERROR": 0,
    "TASK_FAILED": 185,
    "TASK_FINISHED": 14,
    ...
    "hostname": "10.10.12.183",
    "id": "74df965e-a46e-46a9-8d2a-639db0af0e25-S4",
    ...
    "type": "slave",
    ...
  }
]
```

I wonder if there is a way to get more info about the masters so that masters and slaves look more similar? It's too bad this doesn't return info about the masters:

```python
>>> state = client.get_state_summary()
>>> state.keys()
dict_keys(['hostname', 'cluster', 'slaves', 'frameworks'])
>>> state['hostname']
'10.10.12.68'
>>> state['cluster']
'my-cluster'
```

tamarrow commented:
Let's keep the table in the same order, and move `type` to the end of the list.

You can find the mesos id of the leading master with `mesos/master/state.json`, but I'm not aware of a way to get the mesos id of the other masters :/

tamarrow commented:
ok to test

msabramo commented:
OK, I moved the `TYPE` column to the end. And I also added some code that uses a special type for the leading master -- e.g.:

```
$ dcos node
   HOSTNAME         IP                          ID                    TYPE
 10.10.12.174  10.10.12.174  74df965e-a46e-46a9-8d2a-639db0af0e25-S0  slave
 10.10.12.183  10.10.12.183  74df965e-a46e-46a9-8d2a-639db0af0e25-S4  slave
 10.10.12.29   10.10.12.29   74df965e-a46e-46a9-8d2a-639db0af0e25-S2  slave
 10.10.3.201   10.10.3.201   74df965e-a46e-46a9-8d2a-639db0af0e25-S3  slave
master.mesos.  10.10.11.179                    N/A                    master
master.mesos.  10.10.12.68                     N/A                    master (leader)
master.mesos.  10.10.13.72                     N/A                    master
```

tamarrow commented on cli/dcoscli/node/main.py:
```diff
@@ -472,9 +472,21 @@ def _list(json_, extra_field_names):
+        if master['ip'] == master_state['hostname']:
+            master['type'] = 'master (leader)'
+            for key in ('id', 'pid', 'version'):
+                master[key] = master_state.get(key)
```
`--field` is supposed to add fields to the table view that are already in the `json` view. Otherwise we're just hardcoding values and it's not clear what options are available to add to the table. I'm happy to add more info to the json view though (and then it will work more naturally for the table view)

msabramo replied:
This does add it to the `json` view too though.

```
$ dcos node --json | jq '.[0:3]'
[
  {
    "host": "master.mesos.",
    "id": "74df965e-a46e-46a9-8d2a-639db0af0e25",
    "ip": "10.10.12.68",
    "pid": "master@10.10.12.68:5050",
    "type": "master (leader)",
    "version": "1.0.1"
  },
  {
    "host": "master.mesos.",
    "ip": "10.10.11.179",
    "type": "master"
  },
  {
    "host": "master.mesos.",
    "ip": "10.10.13.72",
    "type": "master"
  }
]
```

I think I must be missing your point.

tamarrow replied:
Read this as only adding if `extra_field_names` had those values. please ignore :)

msabramo replied:
Ah okay. Phew!

tamarrow commented on cli/dcoscli/node/main.py:
```diff
@@ -485,7 +497,7 @@ def _list(json_, extra_field_names):
                         'Field "%s" is invalid.' % field_name))
                     return
-        table = tables.slave_table(slaves, extra_field_names)
+        table = tables.node_table(nodes, extra_field_names)
```
we can remove `slave_table` now.

msabramo replied:
500b276

msabramo commented:
Yikes. I'm 0 for 4 on CI on this one.

tamarrow commented:
`test_marathon_debug.test_debug_details` is failing because it uses `dcos node --json` which is now different.

msabramo commented:
It looks to me like that test failed because it only picked up 2 nodes instead of 3. Maybe one was down?

msabramo commented:
but it's hard for me to debug this because I haven't yet figure out how to get the integration tests to work locally for me.

tamarrow commented:
Ah, the test is failing because `dcos node` now has masters + agents, while before it only had agents. So the test needs to be updated.

msabramo commented:
Interesting. My last commit fixed the failing test but now there are a bunch of other failing tests. 
Weird that these didn't show up before. We're things set up so that a failing test would abort the test run before it got to other tests?

I'll work on fixing these failures later when I'm back at a keyboard. I do hope that these are all the test failures, so I won't have to iterate a bunch of times.

msabramo commented:
retest

msabramo commented:
Ah the reason why I'm having to fix the test failures one a time is that the `tox.ini` passed the `-x` option to py.test, which makes it abort on the first test failure. 

I submitted https://github.com/dcos/dcos-cli/pull/964 to remove that option, as it will be helpful to find out about all the test failures at once.

msabramo commented:
I cherry-picked 17679a4 from #964 so I can see what all the failures are.

msabramo commented:
Oh my goodness. Most of the CI tasks passed!

![screen shot 2017-04-05 at 9 54 57 am](https://cloud.githubusercontent.com/assets/305268/24717055/f853565c-19e5-11e7-8ee9-0f65a274126b.png)

msabramo commented:
Here's the `OS X Binary` failure:

```
args = ['--mesos-id=0bd80034-1eac-4d9b-bbf2-d2a6c559c42d-S1', '--master-proxy', '/opt/mesosphere/bin/detect_ip']
expected_returncode = 0, expected_stdout = '10.0.1.38'

    def _node_ssh(args, expected_returncode=None, expected_stdout=None):
        if os.environ.get('CLI_TEST_MASTER_PROXY') and \
                '--master-proxy' not in args:
            args.append('--master-proxy')
    
        stdout, stderr, returncode = _node_ssh_output(args)
>       assert returncode is expected_returncode, \
            'returncode = %r; stdout: = %s; stderr = %s' % (
                returncode, stdout, stderr)
E       AssertionError: returncode = None; stdout: = b'10.0.1.38\r\nConnection to 10.0.1.38 closed.\r\r\n'; stderr = b'Running `ssh -A -t -o StrictHostKeyChecking=no core@52.58.21.89 ssh -A -t -o StrictHostKeyChecking=no core@10.0.1.38 /opt/mesosphere/bin/detect_ip`\nKilled by signal 15.\r\n'
E       assert None is 0

tests/integrations/test_node.py:266: AssertionError
```

msabramo commented:
wat

![screen shot 2017-04-05 at 1 25 03 pm](https://cloud.githubusercontent.com/assets/305268/24725323/52b80f44-1a03-11e7-82ec-0bfd8390568b.png)

tamarrow commented on cli/tests/unit/data/node.txt:
```diff
@@ -1,2 +1,2 @@
-dcos-01   172.17.8.101  20150630-004309-1695027628-5050-1649-S0  
\ No newline at end of file
+HOSTNAME       IP                          ID                    TYPE   
+dcos-01   172.17.8.101  20150630-004309-1695027628-5050-1649-S0  slave  
```
Ah, i overlooked this earlier, for TYPE we should say leader/agent not master/slave. We've been replacing the terminology. Quick change and then lets ship this!

msabramo replied:
Right. `slave` should be `agent`. But I thought `master` was okay because in my mind we need to distinguish between what I think of as multiple masters and the one leader (or leading master). Or have I got the terminology wrong?

msabramo replied:
Also the hostnames still seem to be `master.mesos` so I think it makes sense for the type to be `master` or `master (leader)` like this:

```
$ dcos node
   HOSTNAME         IP                          ID                     TYPE
 10.71.90.121  10.71.90.121  80527015-4faa-4c42-ab6b-7e36c4a22ee5-S0   agent
 10.71.91.112  10.71.91.112  80527015-4faa-4c42-ab6b-7e36c4a22ee5-S5   agent
 10.71.91.137  10.71.91.137  80527015-4faa-4c42-ab6b-7e36c4a22ee5-S12  agent
 10.71.91.25   10.71.91.25   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S4   agent
 10.71.91.37   10.71.91.37   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S1   agent
 10.71.91.49   10.71.91.49   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S11  agent
 10.71.91.71   10.71.91.71   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S2   agent
 10.71.91.73   10.71.91.73   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S7   agent
  10.71.91.8    10.71.91.8   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S14  agent
 10.71.91.84   10.71.91.84   80527015-4faa-4c42-ab6b-7e36c4a22ee5-S3   agent
master.mesos.  10.71.91.155                    N/A                     master
master.mesos.  10.71.91.59                     N/A                     master
master.mesos.  10.71.91.68     f6365617-8938-4968-8421-e80d1ec33143    master (leader)
```

but let me know if you have other ideas.

tamarrow replied:
yep, just confirmed.

msabramo commented:
Grrr. Now *Linux Binary* failed.

msabramo commented:
And now **Linux Binary** passed and **windows-binary** failed. I feel like we're playing CI Roulette here.

msabramo commented:
Yay, all CI tasks passed!  🎉 

![screen shot 2017-04-05 at 11 16 26 pm](https://cloud.githubusercontent.com/assets/305268/24740425/f75dc078-1a55-11e7-9957-c8a054ddb160.png)

tamarrow commented:
🚀
